// @progress/kendo-angular-schematics/src/charts/circular/index.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2023 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.chart = void 0;
const core_1 = require("@angular-devkit/core");
const strings_1 = require("@angular-devkit/core/src/utils/strings");
const schematics_1 = require("@angular-devkit/schematics");
const path_1 = require("path");
const utils_1 = require("../../utils");
const data_1 = require("./files/data");
function chart(_options) {
    return (_tree, _context) => {
        _options.name = (0, path_1.basename)(_options.name);
        let path = (0, path_1.normalize)('/' + (0, path_1.dirname)((_options.path + '/' + _options.name)));
        if (process.platform == "win32") {
            path = path.replace(/\\/g, '/');
        }
        const parsedComplexOptions = (0, utils_1.parseComplexOptions)(['series'], _options);
        const templateSource = (0, schematics_1.apply)((0, schematics_1.url)(_options.inlineTemplate ? './files/inline' : './files/html'), [
            (0, schematics_1.template)({
                classify: core_1.strings.classify,
                dasherize: core_1.strings.dasherize,
                arrayify(args) {
                    return args.map(s => `'${s}'`).join(", ");
                },
                ...Object.assign({}, _options, parsedComplexOptions)
            }),
            (0, schematics_1.move)(_options.path),
        ]);
        const angularComponentOptions = {
            path: _options.path,
            name: _options.name,
            inlineTemplate: _options.inlineTemplate,
            inlineStyle: _options.inlineStyle,
            skipTests: _options.skipTests
        };
        if (_options.mockedData) {
            _tree.create(`${_options.path}/${(0, strings_1.dasherize)(_options.name)}/data.ts`, `export const seriesData = ${data_1.seriesData}`);
        }
        return (0, schematics_1.chain)([
            (0, schematics_1.externalSchematic)('@schematics/angular', 'component', angularComponentOptions),
            (0, schematics_1.mergeWith)(templateSource, schematics_1.MergeStrategy.Overwrite)
        ]);
    };
}
exports.chart = chart;
